### PR TITLE
Support compressed files in data_manager_fetch_genome_all_fasta_dbkeys

### PR DESCRIPTION
**Pull request by mvdbeek**
It's hard these days to find uncompressed fasta downloads anywhere ... so this PR
will check if input files (download or history/directory, except for symlinks for obvious reasons) look like they are compressed and uncompress them before writing them out.

Also minor tweaks to the help text, use of CDATA tag, avoid `<command interpreter="python">`
and py2/py3 compatibility.

I have uploaded the manager under my name to the TTS if someone would like to test https://testtoolshed.g2.bx.psu.edu/view/mvdbeek/data_manager_fetch_genome_dbkeys_all_fasta_compressed/0738889ff4d2 

I couldn't quite figure out if/how I can use planemo to test a data manager, but I have uploaded a workflow [here](https://gist.github.com/mvdbeek/9cee730483a68b36d9bf23c3eb3897af) that tests most of the functionality.

### CODE REVIEW

File: data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py [hunk 1]
```diff
@@ -175,155 +200,163 @@ def _sort_fasta_custom( fasta_filename, params ):
...
+
     ucsc_path = UCSC_DOWNLOAD_PATH % ucsc_dbkey
-    path_contents = _get_files_in_ftp_path( ftp, ucsc_path )
-    
-    ucsc_file_name = None
-    get_stream_reader = None
-    ext = None
-    ucsc_chrom_fa_filename = None
+    path_contents = _get_files_in_ftp_path(ftp, ucsc_path)
+    ftp.close()
```
Comment (nsoranzo) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py]: It seems from Python doc and Google search, that `ftp.quit()` should be the proper way to close the connection, but not really sure.

Comment (mvdbeek) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py]:
```
In [2]: ?FTP.close
Signature: FTP.close(self)
Docstring: Close the connection without assuming anything about it.
File:      /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ftplib.py
Type:      instancemethod

In [3]: ?FTP.quit
Signature: FTP.quit(self)
Docstring: Quit, and close the connection.
File:      /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ftplib.py
Type:      instancemethod
```
Seems quit is more gentle? Will do that.

**Comment by mvdbeek** [comment]
Anyone willing to 👍 ? @bgruening @blankenberg ?

File: data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py [hunk 2]
```diff
@@ -405,11 +439,16 @@ def main():
...
     #Fetch the FASTA
-    REFERENCE_SOURCE_TO_DOWNLOAD[ params['param_dict']['reference_source']['reference_source_selector'] ]( data_manager_dict, params, target_directory, dbkey, dbkey_name, sequence_id, sequence_name )
-    
+    try:
+        REFERENCE_SOURCE_TO_DOWNLOAD[ params['param_dict']['reference_source']['reference_source_selector'] ]( data_manager_dict, params, target_directory, dbkey, dbkey_name, sequence_id, sequence_name, tmp_dir )
+    finally:
+        cleanup_before_exit(tmp_dir)
     #save info to json file
-    open( filename, 'wb' ).write( dumps( data_manager_dict ) )
+    open( filename, 'w' ).write( dumps( data_manager_dict ) )
```
Comment (bgruening) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py]: Why do you change it here and use `wb` above. `wb` does nothing on Unix like operating systems afaik. Just wondering.

File: data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.xml [hunk 3]
```diff
@@ -1,86 +1,85 @@
...
+            </when>
+            <when value="gatk">
+            </when>
+            <when value="custom">
+                <repeat name="sequence_identifiers" title="Sequence Identifiers" min="1" default="1">
+                    <param type="text" name="identifier" value="" label="Sequence Identifier" optional="False" />
+                </repeat>
+                <conditional name="handle_not_listed">
+                    <param name="handle_not_listed_selector" type="select" label="How to handle non-specified Identifiers">
+                        <option value="discard" selected="True">Discard</option>
```
Comment (bgruening) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.xml]: Why is this a conditional? I know it not your fault but maybe we can improve things here a little bit.

File: data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py [hunk 4]
```diff
@@ -175,155 +200,163 @@ def _sort_fasta_custom( fasta_filename, params ):
...
+
+def _get_ucsc_download_address(params, dbkey):
+    """
+    Check if we can find the correct file for the supplied dbkey on UCSC's FTP server
+    """
     UCSC_FTP_SERVER = 'hgdownload.cse.ucsc.edu'
     UCSC_DOWNLOAD_PATH = '/goldenPath/%s/bigZips/'
-    COMPRESSED_EXTENSIONS = [ ( '.tar.gz', _get_stream_readers_for_tar ), ( '.tar.bz2', _get_stream_readers_for_tar ), ( '.zip', _get_stream_readers_for_zip ), ( '.fa.gz', _get_stream_readers_for_gzip ), ( '.fa.bz2', _get_stream_readers_for_bz2 ) ]
-    
+    COMPRESSED_EXTENSIONS = ['.tar.gz', '.tar.bz2', '.zip', '.fa.gz', '.fa.bz2']
```
Comment (bgruening) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py]: tgz?

File: data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py (hunk 2 above)
Comment (mvdbeek) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py]: I'll check that. IIRC python3 didn't want to dump the json file when the file was opened in 'wb' mode.

File: data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.xml (hunk 3 above)
Comment (mvdbeek) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.xml]: You're right, I think we can simply remove the conditional.

File: data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py (hunk 4 above)
Comment (mvdbeek) [inline data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py]: Safe addition, not sure they really have that extension, but I'll add it.

**Comment by mvdbeek** [comment]
@bgruening I removed the conditional and added .tgz to the list of files downloadable from UCSC.
About the binary mode: I played around with this some more, python3 will refuse to write bytes (such as received by downloading files or reading in intermediary fasta files) to files open only in `'w'` mode. So I also tried decoding the incoming bytestreams and encoding before feeding this back to the compressed stream handler, but there was always something wrong in these modules in conjunction with StringIO or BytesIO.
I think this is currently the easiest solution to support both python3 and python2 (maybe not the best, but I don't think this would be a regression from the previous code).

**Comment by bgruening** [comment]
Sounds good to me! Thanks!

**Comment by mvdbeek** [comment]
Thanks @bgruening ! @martenson can you push it pleaseee?

**Comment by martenson** [comment]
@mvdbeek done https://toolshed.g2.bx.psu.edu/view/devteam/data_manager_fetch_genome_dbkeys_all_fasta/86fa71e9b427

**Comment by mvdbeek** [comment]
Thanks @martenson

**Pull request merged**